innkeeper/client_test.go

```diff
@@ -0,0 +1,601 @@
+package innkeeper
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/zalando/eskip"
+	"github.com/zalando/skipper/skipper"
+	"net/http"
+	"net/http/httptest"
+	"path"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+)
+
+const testAuthenticationToken = "test token"
+
+type autoAuth bool
+
+func (aa autoAuth) Token() (string, error) {
+	if aa {
+		return testAuthenticationToken, nil
+	}
+
+	return "", errors.New(string(authErrorAuthentication))
+}
+
+type innkeeperHandler struct{ data []*routeData }
+
+func (h *innkeeperHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Header.Get(authHeaderName) != testAuthenticationToken {
+		w.WriteHeader(http.StatusUnauthorized)
+		enc := json.NewEncoder(w)
+
+		// ignoring error
+		enc.Encode(&apiError{ErrorType: string(authErrorPermission)})
+
+		return
+	}
+
+	var responseData []*routeData
+	if r.URL.Path == "/routes" {
+		for _, di := range h.data {
+			if di.DeletedAt == "" {
+				responseData = append(responseData, di)
+			}
+		}
+	} else {
+		lm := path.Base(r.URL.Path)
+		if lm == updatePathRoot {
+			lm = ""
+		}
+
+		responseData = []*routeData{}
+		for _, di := range h.data {
+			if di.CreatedAt > lm || di.DeletedAt > lm {
+				responseData = append(responseData, di)
+			}
+		}
+	}
+
+	if b, err := json.Marshal(responseData); err == nil {
+		w.Write(b)
+	} else {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func sortDoc(doc string) string {
+	exps := strings.Split(doc, ";")
+	for i, exp := range exps {
+		exps[i] = strings.Trim(exp, " \n")
+	}
+	sort.Strings(exps)
+	return strings.Join(exps, ";")
+}
+
+func checkDoc(out skipper.RawData, in []*routeData) bool {
+	c := &Client{doc: make(map[int64]map[string]string)}
```

Comment:
Should we define a type alias for this? I've seen it somewhere else. Maybe this is my C-troll talking...